contracts/sol4/bridgeReserves/uniswapCurve/KyberUniswapCurveReserve.sol

```diff
@@ -0,0 +1,466 @@
+            // first, not use Curve, get amount eth-> dest in Uniswap
+            destAmount = getUniswapDestAmount(dest, srcQty, true);
+            useCurve = false;
+            tokens = bridgeTokens[dest];
+            for(i = 0; i < tokens.length; i++) {
+                token = ERC20(tokens[i]);
+                // swap eth -> token in Uniswap, token -> dest in Curve
+                destQtyBridgeToken = getUniswapDestAmount(token, srcQty, true);
+                if (destQtyBridgeToken > 0) {
+                    destQtyBridgeToken = getCurveDestAmount(token, dest, destQtyBridgeToken);
+                    if (destQtyBridgeToken > destAmount) {
+                        destAmount = destQtyBridgeToken;
+                        bridgePosition = i;
+                        bridgeToken = token;
+                        useCurve = true;
+                    }
+                }
+            }
+        } else {
+            // check src -> token in Curve, token -> eth in Uniswap
+            // first try to not use Curve
+            destAmount = getUniswapDestAmount(src, srcQty, false);
+            useCurve = false;
+            tokens = bridgeTokens[src];
+            for(i = 0; i < tokens.length; i++) {
+                token = ERC20(tokens[i]);
+                // swap src -> token in Curve, token -> eth in Uniswap
+                destQtyBridgeToken = getCurveDestAmount(src, token, srcQty);
+                if (destQtyBridgeToken > 0) {
+                    destQtyBridgeToken = getUniswapDestAmount(token, destQtyBridgeToken, false);
+                    if (destQtyBridgeToken > destAmount) {
+                        destAmount = destQtyBridgeToken;
+                        bridgePosition = i;
+                        bridgeToken = token;
+                        useCurve = true;
+                    }
+                }
+            }
+        }
+    }
+
+    function doTradeEthToToken(
+        ERC20 token,
+        ERC20 bridgeToken,
+        bool useCurve,
+        uint256 srcAmount
+    )
+        internal returns(uint destAmount)
+    {
+        uint256 balanceBefore = token.balanceOf(address(this));
+        address[] memory path = new address[](2);
+        path[0] = weth;
+        if (!useCurve) {
+            // directly swap with Uniswap
+            path[1] = address(token);
+            uniswapRouter.swapExactETHForTokens.value(srcAmount)(
+                0, path, address(this), DEADLINE
+            );
+        } else {
+            // swap eth -> bridge token on Uniswap
+            uint256 bridgeTokenBalBefore = bridgeToken.balanceOf(address(this));
+            path[1] = address(bridgeToken);
+            uniswapRouter.swapExactETHForTokens.value(srcAmount)(
+                0, path, address(this), DEADLINE
+            );
+            uint256 bridgeTokenBalAfter = bridgeToken.balanceOf(address(this));
+            require(bridgeTokenBalAfter >= bridgeTokenBalBefore);
```

Comment:
assert?
not sure we need this
do we suspect uniswap not to return tokens?
it seems very un likely.
so not sure we should keep the balance check before / after uniswap